watcher.go

```diff
@@ -0,0 +1,67 @@
+package consul
+
+import (
+	"context"
+	"strconv"
+)
+
+type WatcherFunc func([]KeyData)
+
+type Watcher struct {
+	Client *Client
+}
+
+var (
+	DefaultWatcher = &Watcher{Client: DefaultClient}
+)
+
+func Watch(ctx context.Context, key string, handler WatcherFunc) {
+	DefaultWatcher.Watch(ctx, key, handler)
+}
+
+func (w *Watcher) Watch(ctx context.Context, key string, handler WatcherFunc) {
+	w.watching(ctx, key, handler, nil)
+}
+
+func WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	DefaultWatcher.WatchPrefix(ctx, prefix, handler)
+}
+
+func (w *Watcher) WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	q := Query{Param{
+		Name:  "recurse",
+		Value: "",
+	}}
+	w.watching(ctx, prefix, handler, q)
+}
+
+func updateParam(q Query, new Param) Query {
+	ret := make(Query, len(q))
+	copy(ret, q)
+	for i, p := range q {
+		if new.Name == p.Name {
+			ret[i] = new
+			return ret
+		}
+	}
+	return append(ret, new)
+}
+
+func (w *Watcher) watching(ctx context.Context, key string, handler WatcherFunc, q Query) {
+	q = updateParam(q, Param{Name: "index", Value: "0"})
+	resp := []KeyData{}
+	path := "/v1/kv/" + key
+	hdr, err := w.Client.do(ctx, "GET", path, q, nil, &resp)
+	if err != nil {
+		// What to do?
+	}
+	for {
+		q = updateParam(q, Param{Name: "index", Value: strconv.FormatUint(hdr.index, 10)})
+		resp := make([]KeyData, 0)
+		hdr, err = w.Client.do(ctx, "GET", path, q, nil, &resp)
+		if err != nil {
+			// What to do?
```

Comment:
Should we change the handler's function signature to receive the error `func([]KeyData, error)`?

Then we must check for `ctx.Err() != nil` and return in that case so the function eventually returns when the context has been canceled.

Reply:
This is tricky. What is the caller going to do with the error?  They can't easily re-execute the request because they don't access to the scope.

The errors I see likely to happen are either network related or the API becoming unavailable.  I think we should add retries for things that can be done, otherwise terminate and return an error from watch/watchprefix.  Thoughts?

Reply:
I think passing the error will allow the caller to log or report it through other means, potentially cancelling the context to terminate the watcher if it wishes. Just swallowing it and not letting the handler know that something went wrong is a risky bet because it could appear as "not making progress" and blocking forever.

Reply:
sgtm.  How about we do this.  Check if it's a temporary error and retry if so.  If it's not, then send the error on the handler and backoff, allowing the caller to make the decision you described.

Reply:
Sounds alright 👍